### PR TITLE
Virtual DOM APIs in Landing 

### LINKED ISSUE #823: Make landing page/launcher use the virtual DOM
This is a perfect widget to use the virtual DOM with.

### PR DESCRIPTION
**Pull request by charnpreetsingh**
Closes #823

### CODE REVIEW

**Comment by sccolbert** [comment]
This is a lot of logic shoved into the `activate` function. Any reason why this rendering logic is not encapsulated in its own widget?

**Comment by charnpreetsingh185** [comment]
@sccolbert No particular reason. I'll move it into the `LandingWidget` class itself.

**Comment by ellisonbg** [comment]
Hi @charnpreetsingh185 !

To help simplify the patterns of working with the virtual DOM APIS, we have created a new `VDomModel` and `VDomWidget` classes that encode the best practices for working with this API. These new APIs are documented here:

https://github.com/jupyter/jupyterlab/blob/master/tutorial/virtualdom.md

Could you update this PR to use these classes? Please let me know if you have any questions!

**Comment by charnpreetsingh185** [comment]
@ellisonbg Yes, I saw that PR I will update this ASAP.

**Comment by charnpreetsingh185** [comment]
`VDomModel` and `VDomWidget` classes added.

**Comment by ellisonbg** [comment]
@charnpreetsingh185 can you rebase?

**Comment by charnpreetsingh185** [comment]
@ellisonbg rebased.

**Comment by ellisonbg** [comment]
Thanks I will review...

**Comment by ellisonbg** [comment]
Grr, looks like you need to rebase again...sorry about this...

File: src/landing/plugin.ts [hunk 1]
```diff
@@ -32,107 +36,113 @@ const landingExtension: JupyterLabPlugin<void> = {
...
+  set path(value: string) {
+    this._path = value;
+    this.stateChanged.emit(void 0);
   }
 
-  let img = body.getElementsByClassName('jp-ImageNotebook')[0];
-  img.addEventListener('click', () => {
-    app.commands.execute('file-operations:new-notebook', void 0);
-  });
+  get data(): VNode[] {
```
Comment (ellisonbg) [inline src/landing/plugin.ts]: In general, the model should only have state related things and shouldn't know about the virtual DOM in any way. I would move the code that builds vdom elements using the actual model data to the VDomWidget subclass.

File: src/landing/plugin.ts [hunk 2]
```diff
@@ -32,107 +36,113 @@ const landingExtension: JupyterLabPlugin<void> = {
...
 
-  img = body.getElementsByClassName('jp-ImageTextEditor')[0];
-  img.addEventListener('click', () => {
-    app.commands.execute('file-operations:new-text-file', void 0);
-  });
+  constructor(app: JupyterLab, services: IServiceManager) {
+    super();
+    let previewMessages = ['super alpha preview', 'very alpha preview', 'extremely alpha preview', 'exceedingly alpha preview', 'alpha alpha preview'];
+    let actualMessage = previewMessages[(Math.floor(Math.random() * previewMessages.length))];
+    let activitiesList: VNode[] = [];
```
Comment (ellisonbg) [inline src/landing/plugin.ts]: Move this vdom stuff to the widget.

File: src/landing/plugin.ts [hunk 3]
```diff
@@ -32,107 +36,113 @@ const landingExtension: JupyterLabPlugin<void> = {
...
+    const activites =
+      [['Notebook', 'file-operations:new-notebook'],
+       ['Code Console', `console:create-${services.kernelspecs.default}`],
+       ['Terminal', 'terminal:create-new'],
+       ['Text Editor', 'file-operations:new-text-file']];
+    for (let activityName of activites) {
```
Comment (ellisonbg) [inline src/landing/plugin.ts]: All of this vdom stuff to the widget

File: src/landing/plugin.ts [hunk 4]
```diff
@@ -32,107 +36,113 @@ const landingExtension: JupyterLabPlugin<void> = {
...
-  cwd.className = 'jp-Landing-cwd';
+  private _path: string;
+  private _data: VNode[];
+}
 
+class LandingWidget extends VDomWidget<LandingModel> {
+  protected render(): VNode {
+    let dialog =
+    h.div({className: 'jp-Landing-dialog'},
+      this.model.data,
```
Comment (ellisonbg) [inline src/landing/plugin.ts]: Replace `this.model.data` with inline generation of those vdom elements...

**Comment by charnpreetsingh185** [comment]
Rebased. Working on making the changes requested.

**Comment by charnpreetsingh185** [comment]
@ellisonbg Changes made

**Comment by ellisonbg** [comment]
This is getting close! There is still some refactoring needed to better separate the model/view stuff - see inline comments.

**Comment by charnpreetsingh185** [comment]
@ellisonbg This is ready for another review pass.

**Comment by blink1073** [comment]
Nice, thanks!

**Pull request merged**